### DISCUSSION TITLE
Add support for using ssh agent

### DISCUSSION THREAD

**Discussion by tzeumer** [doublecmd/doublecmd#1133] [Suggestions]
### Description

I nearly didn't click on "Network > FTP" because I didn't expect an SSH option. Anyway, great that ssh connections are possible. 

Still, it would be great if there also were an option to use the/an ssh agent.

**Comment by masscream** [comment]
Since FileZilla declined to make a support for specifying a keypair with the use of ssh-agent - [here](https://trac.filezilla-project.org/ticket/8450), and never made any alternative custom setting ([here](https://trac.filezilla-project.org/ticket/7739)). It would be great to have a modern looking gui sftp file manager for Linux. Yes, there is still gftp to use.... but.... I totaly agree this option would be awesome.

**Comment by alexx2000** [comment]
Added (revision 11046).

**Comment by njeno94** [comment]
Hi @alexx2000 !

Is it possible to use ssh-agent, when I open an sftp location with the folder path like: sftp://username@hostname?
It would be great, because in this case the .ssh/config parsed so we can use ssh certificates and jumphosts to reach the target server.

**Comment by alexx2000** [comment]
Hi, when you open path like `sftp://username@hostname` then DC use GVFS. So you should find how to enable ssh-agent for GVFS.

**Comment by njeno94** [comment]
Thx, it works now on Linux!

I also tested on windows but in that version I cannot see the Use SSH-agent option in the FTP connection parameters. (GVFS doesn't work at all)
Is it possible to implement in the FTP plugin to use the .ssh/config file for scp/sftp connections?

**Comment by masscream** [comment]
Thanks, I've tested this as successful. Except one of the devices is still getting "maximum authentication attempts exceeded". The agent authentication truly works, but is it respecting .ssh/config file ? That was actually the original idea, why I mentioned it here.... and also a reason why ssh-agent and/or FileZilla are useless tools together so far... and I have also found no suitable solution for this. Except gFTP, which is a mess.

**Comment by alexx2000** [comment]
> The agent authentication truly works, but is it respecting .ssh/config file ?

I guess no. Agent only gives a list of public keys and capability to try each key. 

https://libssh2.org/examples/ssh2_agent.html

**Comment by masscream** [comment]
Could that be implemented please ? I really wonder how ssh-agent could be usable without specifying the right key.... with the growing number of servers, you just would have to allow brute-forcing them with bunch of keys... in case you even could.

I switched from keys to agent with the motivation of higher security, if there is not a suitable solution for files transfer, I guess I'll dump it back cause the agent is a nice try but bad result...

**Comment by alexx2000** [comment]
I'll think about it, it's not so easy. This will require creating a `.ssh/config` parser.

**Comment by njeno94** [comment]
Did you think about changing the libssh2 library to libssh?  It already supports parsing the ssh config file and has other features.

**Comment by alexx2000** [comment]
No, it is too hard.

**Comment by Robert-M-Muench** [comment]
I'm on macOS using 1Password as my ssh-agent.

So, do I understand it correctly that it's not possible to use ssh-agent for SFTP connection at the moment?

Is it at all possible to use an ssh key?

**Reply by alexx2000** [reply]
DC supports both methods: ssh-agent and ssh-keys.

**Reply by Robert-M-Muench** [reply]
Doesn't seem to work with 1Password on macOS:

```
Info: [3], Connecting to: ssh.abc.def.host
Info: [3], ssh.abc.def.host:22
Info: [3], 53.147.171.12:22
Info: [3], Key exchange method: curve25519-sha256
Info: [3], Encryption method: chacha20-poly1305@openssh.com
Info: [3], Host key method: rsa-sha2-512
Info: [1], 
Info: [3], Connection established
Info: [3], Server fingerprint: (MD5) DB 70 8A 20 96 EA 88 AE B2 D4 FB 93 BB 0D C0 F0
Info: [3], Server fingerprint: (SHA1) 13 EF 3E B3 72 3E AB 1A D8 AA 36 8A 06 43 F0 1D 93 74 89 66
Info: [3], Server fingerprint: (SHA256) 1t4Gslc+NvEHX0ufDbgLkiPkmzGfWWxDLJXd8Ilp0Wo
Info: [3], Authentication methods: password,publickey
Info: [3], SSH-agent authentication
Error: [6], SSH ERROR            
```
No SSH agent pops up, as it should with 1P.
